fdbclient/include/fdbclient/MetaclusterManagement.actor.h

```diff
@@ -199,9 +200,29 @@ struct MetaclusterOperationContext {
 
 	Optional<MetaclusterRegistrationEntry> metaclusterRegistration;
 	Optional<DataClusterMetadata> dataClusterMetadata;
+	bool dataClusterIsRegistered = true;
+
+	std::set<DataClusterState> extraSupportedDataClusterStates;
+
+	MetaclusterOperationContext(Reference<DB> managementDb,
+	                            Optional<ClusterName> clusterName = {},
+	                            std::set<DataClusterState> extraSupportedDataClusterStates = {})
+	  : managementDb(managementDb), clusterName(clusterName),
+	    extraSupportedDataClusterStates(extraSupportedDataClusterStates) {}
+
+	void checkClusterState() {
+		DataClusterState clusterState =
+		    dataClusterMetadata.present() ? dataClusterMetadata.get().entry.clusterState : DataClusterState::READY;
+		if (clusterState != DataClusterState::READY && extraSupportedDataClusterStates.count(clusterState) == 0) {
+			if (clusterState == DataClusterState::REGISTERING) {
+				throw cluster_not_found();
+			} else if (clusterState == DataClusterState::REMOVING) {
+				throw cluster_removed();
+			}
 
-	MetaclusterOperationContext(Reference<DB> managementDb, Optional<ClusterName> clusterName = {})
-	  : managementDb(managementDb), clusterName(clusterName) {}
+			ASSERT(false);
```

Comment:
Is it possible for the cluster state to be `DataClusterState::RESTORING` when we do this check? Is there any workload test sequence that would lead us to that situation?

Reply:
Not currently, but this will need to be handled in the PR that adds support for restoring clusters in a metacluster that I have open.